ingestion/src/kvkafka.rs

```diff
@@ -179,54 +193,112 @@ fn parse_message(message: &str) -> Result<Vec<RawDatum>, Error> {
+    transaction
+        .execute(
+            "LOCK TABLE public.timeseries IN SHARE ROW EXCLUSIVE MODE",
+            &[],
+        )
+        .await?;
+
+    // re-check for an existing label since the first check was outside the transaction
+    let rows = transaction
+        .query(
+            QUERY_GET_MET_STR,
+            &[
+                &raw_datum.kvid.station,
+                &raw_datum.kvid.paramid,
+                &raw_datum.kvid.typeid,
+                &raw_datum.kvid.level,
+                &raw_datum.kvid.sensor,
+            ],
+        )
+        .await?;
+    if let Some(row) = rows.first() {
+        return Ok(row.get(0));
+    }
 
-    let mut open_data: Vec<Datum> = Vec::new();
-    let mut restricted_data: Vec<Datum> = Vec::new();
-
-    // TODO: transforming this to an iterator would let us pipeline the queries
-    for raw_datum in raw_data {
-        let permit = timeseries_get_permit(
-            permit_table.clone(),
-            raw_datum.kvid.station,
-            raw_datum.kvid.typeid,
-            raw_datum.kvid.paramid,
-        )?;
-
-        let (transaction, query_met, data) = match permit {
-            Some(1) => (
-                open_conn.transaction().await?,
-                &query_met_open,
-                &mut open_data,
-            ),
-            _ => (
-                restricted_conn.transaction().await?,
-                &query_met_restricted,
-                &mut restricted_data,
-            ),
-        };
+    // TODO: currently we create a timeseries with null location
+    // In the future the location column should be moved to the timeseries metadata table
+    let timeseries_id = transaction
+        .query_one(
+            "INSERT INTO public.timeseries (fromtime, permit) VALUES ($1, $2) RETURNING id",
+            &[&raw_datum.obstime, &permit],
+        )
+        .await?
+        .get(0);
+
+    // create source-specific label
+    transaction
+        .execute(
+            "INSERT INTO labels.kvalobs \
+        (timeseries, station_id, param_id, type_id, lvl, sensor) \
+    VALUES ($1, $2, $3, $4, $5, $6)",
+            &[
+                &timeseries_id,
+                &raw_datum.kvid.station,
+                &raw_datum.kvid.paramid,
+                &raw_datum.kvid.typeid,
+                &raw_datum.kvid.level,
+                &raw_datum.kvid.sensor,
+            ],
+        )
+        .await?;
+
+    // create met label
+    transaction
+        .execute(
+            "INSERT INTO labels.met \
+        (timeseries, station_id, param_id, type_id, lvl, sensor) \
+    VALUES ($1, $2, $3, $4, $5, $6)",
+            &[
+                &timeseries_id,
+                &raw_datum.kvid.station,
+                &raw_datum.kvid.paramid,
+                &raw_datum.kvid.typeid,
+                &raw_datum.kvid.level,
+                &raw_datum.kvid.sensor,
+            ],
+        )
+        .await?;
+
+    transaction.commit().await?;
```

Comment:
Do you need to explicitly unlock? or is it implicit after the transaction is commited / completed?

Reply:
Yes, locks only apply to a single transaction, and release when it completes. Postgres doesn't provide any way to explicitly release a lock other than committing or rolling back the transaction holding them.

Reply:
Ok, maybe just put a comment about how this releases the lock (for posterity / future people's understanding :)

Reply:
Done